package generator;

import com.beust.jcommander.JCommander;
import com.beust.jcommander.Parameter;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLMapper;
import common.CommonFunctions;
import model.ContactData;
import model.GroupData;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;

import static common.CommonFunctions.randomString;

public class Generator {

    @Parameter(names = {"--type", "-t"})
    String type;

    @Parameter(names = {"--output", "-o"})
    String output;

    @Parameter(names = {"--format", "-f"})
    String format;

    @Parameter(names = {"--count", "-c"})
    int count;

    public static void main(String[] args) throws IOException {
        var generator = new Generator();
        JCommander.newBuilder()
                .addObject(generator)
                .build().parse(args);
        System.out.println("type = " + generator.type);
        System.out.println("output = " + generator.output);
        System.out.println("format = " + generator.format);
        System.out.println("count = " + generator.count);
        generator.run();
    }

    private void run()throws IOException {
        var data = generate();
        save(data);
    }

    private void save(Object data) throws IOException {
        if ("json".equals(format)) {
            ObjectMapper mapper = new ObjectMapper();
            mapper.writeValue(new File(output), data);
        }else if ("yaml".equals(format)){
            ObjectMapper mapper = new YAMLMapper();
            mapper.writeValue(new File(output), data);
        }
        else if ("xml".equals(format)) {
        ObjectMapper mapper = new XmlMapper();
            mapper.writeValue(new File(output), data);
        } else {
            throw new IllegalArgumentException("Неизвестный формат данных" + format);
        }
}

    private Object generate() {
        System.out.println(type);
    if ("groups".equals(type)) {
            return generateGroups();
        } else if ("contacts".equals(type)) {
            return generateContacts();
        } else {
            throw new IllegalArgumentException("Неизвестный тип данных" + type);
        }
    }

    private Object generateGroups() {
        var result = new ArrayList<GroupData>();
        for (int i = 0; i < count; i++) {
            result.add(new GroupData().withName(CommonFunctions.randomString(i * 4)).withHeader(CommonFunctions.randomString(i * 4)).withFooter(randomString(i * 4)));
        }
        return result;
    }

    private Object generateContacts() {
        var result = new ArrayList<ContactData>();
        for (int i = 0; i < count; i++){
            result.add(new ContactData().withFioAndNumber(CommonFunctions.randomString(i*3), CommonFunctions.randomString(i*2),CommonFunctions.randomString(i*4),CommonFunctions.randomString(i*3)));
        }
        return result;
    }
}


